lib/Sema/SemaChecking.cpp

```diff
@@ -10621,6 +10621,73 @@ void Sema::DiagnoseSelfMove(const Expr *LHSExpr, const Expr *RHSExpr,
                                         << RHSExpr->getSourceRange();
 }
 
+
+//===--- CHECK: Checked scope -------------------------===//
+void Sema::DiagnoseCheckedDecl(const DeclStmt *DS) {
+  // Checked C - check consistency between compound stmt checked property
+  // and declaration type
+  DeclGroupRef DG = DS->getDeclGroup();
+  for (DeclGroupRef::const_iterator D = DG.begin(), DEnd = DG.end(); D != DEnd;
+       ++D) {
+    if (const VarDecl *VD = dyn_cast<VarDecl>(*D)) {
+      DiagnoseCheckedDecl(VD);
+    }
+  }
+}
+
+// type restrictions on declaration in checked blocks
+void Sema::DiagnoseCheckedDecl(const DeclaratorDecl *Decl) {
+  // Checked C - check consistency between checked property and declaration
+  // checked pointer type or unchecked pointer type with bounds-safe interface
+  // is only allowed in checked scope or funcion
+  const DeclaratorDecl *TargetDecl = nullptr;
+  int DeclKind;
+  QualType Ty;
+  if (const ParmVarDecl *Parm = dyn_cast<ParmVarDecl>(Decl)) {
+    TargetDecl = Parm;
+    DeclKind = 0; // function param
+    Ty = Parm->getType();
+  }
+  else if (const FunctionDecl *Func = dyn_cast<FunctionDecl>(Decl)) {
+    TargetDecl = Func;
+    DeclKind = 1; // function return
+    Ty = Func->getReturnType();
+  }
+  else if (const VarDecl *Var = dyn_cast<VarDecl>(Decl)) {
+    TargetDecl = Var;
+    DeclKind = 2; // decl var
+    Ty = Var->getType();
+  }
+  else if (const FieldDecl *Field = dyn_cast<FieldDecl>(Decl)) {
+    TargetDecl = Field;
+    DeclKind = 3; // member
+    Ty = Field->getType();
+  }
+
+  bool checkedScopeError = false;
+  int TypeKind = 0;
+  if (TargetDecl) {
+    // type is unchecked pointer w/o bounds-safe interface
+    // type is unchecked pointer & its bounds-safe interface is also unchecked
+    QualType InterOpTy = GetCheckedCInteropType(TargetDecl);
+    if ((Ty->isUncheckedPointerType() || Ty->isUncheckedArrayType()) &&
+        (InterOpTy.isNull() || (InterOpTy->isUncheckedPointerType() ||
+                                InterOpTy->isUncheckedArrayType()))) {
+      checkedScopeError = true;
+      TypeKind =
```

Comment:
I have missed that interoperation type (itype) SHOULD require checked type
If interop type is not null, it is definitely checked type.
So, it is not necessary to check if interop type is checked or unchecked
I will simplify above codes